db/commitlog/commitlog.cc

```diff
@@ -2675,6 +2689,13 @@ db::commitlog::read_log_file(sstring filename, sstring pfx, commit_load_reader_f
                 // if a chunk header checksum is broken, we shall just assume that all
                 // remaining is as well. We cannot trust the "next" pointer, so...
                 clogger.debug("Checksum error in segment chunk at {}.", start);
+
+                if (corrupt_size == 0) {
+                    // if we got here and had no broken data previously, we can 
+                    // just call it truncation
```

Comment:
But how can we distinguish truncation vs. corruption in the middle of the commitlog file?
That fact that this was found first doesn't mean nothing valid was written after it.
Only with sector headers we can validate that the tail of the reused segment consists of valid sectors from previous generations.

Reply:
Don't misinterpret the "just call it" as a de-escalation. Truncation is worse than corruption, because we cannot continue. Truncation can be the result of corruption, but the salient point is that we cannot determine how much was borked, nor recover (no way to safely skip). 
I am perhaps less sure about the `corrupt_size != 0` case - here we try to explicitly add to corrupt size, even though we can't really determine it - because we can guess that if we have broken data in otherwise valid chunks and then run into garbage here, there are at least odds of the cause. 
But no, this is not certain. The only sort of certain part is separating the "no corruption - but unreadable chunk" into its own case. 
Again, not trying to downplay it. But also trying to be more precise of the probable cause of data loss.

Reply:
Wouldn't that be likely to happen on the last segment in unclean shutdown?
How is it avoided in clean shutdown? by having no replay_position beyond the valid bytes after successful memtable flush?

Reply:
Clean shutdown all segments are deleted. If you mean terminated segment, we write a zero-page to the end (or we fill the full segment -> EOF).